lib/bundler/definition.rb

```diff
@@ -256,9 +256,39 @@ def lock(file)
         "#{File.expand_path(file)}"
     end
 
+    # Returns the version of Bundler that is creating or has created
+    # Gemfile.lock. Used in #to_lock.
+    def lock_version
+      lockfile = @lockfile_contents
+      lock_ver = nil
+      curr_ver = nil
+
+      if lockfile and lockfile.match(/^  \[(.*)\]$/)
+        lock_ver = $1
+        curr_ver = Bundler::VERSION
+      end
+
+      if lock_ver and curr_ver
+        if Gem::Version.new(curr_ver) >= Gem::Version.new(lock_ver)
+          ver = curr_ver
+        else
+          ver = lock_ver
+        end
+      else
+        ver = Bundler::VERSION
+      end
+
+      ver
+    end
```

Comment:
So I kept mulling this over, and I think I see some ways to cut down on the amount of code that exists without making it (much?) more obscure:

```ruby
lock_ver = @lockfile_contents[/^  \[(.*)\]$/, 1] if lockfile

if lock_ver && Gem::Version.new(lock_ver) < Gem::Version.new(Bundler::VERSION)
  new_ver = Bundler::VERSION
end

new_ver || lock_ver || Bundler::VERSION
```